File: JavaComparisonModule/ContactBookJava/src/main/java/com/github/insanusmokrassar/KotlinVSJava/JavaComparisonModule/ContactBookJava/ContactBook.java
package com.github.insanusmokrassar.KotlinVSJava.JavaComparisonModule.ContactBookJava;

import com.github.insanusmokrassar.KotlinVSJava.JavaComparisonModule.ContactBookJava.API.Contact;
import com.github.insanusmokrassar.KotlinVSJava.JavaComparisonModule.ContactBookJava.API.ContactsBook;
import com.github.insanusmokrassar.KotlinVSJava.report_formatter.LogPoint;

import java.util.ArrayList;
import java.util.List;

import static com.github.insanusmokrassar.KotlinVSJava.report_formatter.LogPointKt.*;
import static com.github.insanusmokrassar.KotlinVSJava.report_formatter.TestDataFormatterKt.printPoints;

public class ContactBook {
    private final static Integer defaultBooksCount = 10;

    private final static Integer defaultContactsPerBook = 5000;

    private final static String testName = "ContactsBook.Java";

    public static void main(String[] args) {
        Integer booksCount = defaultBooksCount;

        if (args.length > 0) {
            try {
                booksCount = Integer.valueOf(args[0]);
            } catch (NumberFormatException ignored) {}
        }

        Integer contactsPerBook = defaultContactsPerBook;

        if (args.length > 1) {
            try {
                contactsPerBook = Integer.valueOf(args[1]);
            } catch (NumberFormatException ignored) {}
        }

        List<LogPoint> points = new ArrayList<>();

        points.add(new LogPoint(testName, START_TEST));

        List<ContactsBook> books = new ArrayList<>();

        for (int it = 0; it < booksCount; it++) {
            ContactsBook book = new ContactsBook();

            for (int i = 0; i < contactsPerBook; i++) {
                int index = (i * 2) - it;

                book.add(
                        new Contact(
                                "Name " + index,
                                "Last Name " + index,
                                (index + 100) + "",
                                index + "@main.hello"
                        )
                );
            }

            books.add(book);
        }

        points.add(new LogPoint(testName, CONTACTS_BOOKS_FILLED));

        for (ContactsBook mainBook : books) {

            for (int i = 0; i < mainBook.size(); i++) {
                Integer searchIndex = i * 2;
                Contact contact = mainBook.get(i);

                for (ContactsBook book : books) {

                    if (!book.equals(mainBook)) {
                        if (book.indexOf(contact) > searchIndex) {
                            break;
                        }
                    }
                }
            }
        }

        points.add(new LogPoint(testName, CONTACTS_BOOKS_SEARCH_COMPLETED));
        points.add(new LogPoint(testName, COMPLETE_TEST));

        printPoints(points);
    }
}

File: kotlin/ContactBookKotlin/src/main/kotlin/com/github/insanusmokrassar/KotlinVSJava/KotlinComparisonModule/ContactBookKotlin/ContactBook.kt
package com.github.insanusmokrassar.KotlinVSJava.KotlinComparisonModule.ContactBookKotlin

import com.github.insanusmokrassar.KotlinVSJava.KotlinComparisonModule.ContactBookKotlin.API.Contact
import com.github.insanusmokrassar.KotlinVSJava.KotlinComparisonModule.ContactBookKotlin.API.ContactsBook
import com.github.insanusmokrassar.KotlinVSJava.report_formatter.*

private const val defaultBooksCount = 10

private const val defaultContactsPerBook = 5000

private const val testName = "ContactsBook.Kotlin"

fun main(args: Array<String>) {
    val booksCount = if (args.isNotEmpty()) {
        args[0].toIntOrNull() ?: defaultBooksCount
    } else {
        defaultBooksCount
    }
    val contactsPerBook = if (args.size > 1) {
        args[1].toIntOrNull() ?: defaultContactsPerBook
    } else {
        defaultContactsPerBook
    }

    val points = ArrayList<LogPoint>()

    points.add(LogPoint(testName, START_TEST))

    (0 until booksCount).map {
        val contactBook = ContactsBook()

        for (i in (0 until contactsPerBook)) {
            val index = (i * 2) - it
            contactBook.add(
                Contact(
                    "Name $index",
                    "Last Name $index",
                    "${index + 100}",
                    "$index@mail.hello"
                )
            )
        }

        contactBook
    }.let {
        books ->
        points.add(LogPoint(testName, CONTACTS_BOOKS_FILLED))
        books.firstOrNull() ?.let {
            mainBook ->
            mainBook.forEachIndexed {
                i, contact ->
                for (book in books) {
                    if (book != mainBook) {
                        if (book.indexOf(contact) > i) {
                            break
                        }
                    }
                }
            }
        }
    }
    points.add(LogPoint(testName, CONTACTS_BOOKS_SEARCH_COMPLETED))
    points.add(LogPoint(testName, COMPLETE_TEST))

    printPoints(points)
}
